client/js/src/helpers/data.ts

```diff
@@ -118,3 +120,61 @@ export function post_message<Res = any>(
 		window.parent.postMessage(message, origin, [channel.port2]);
 	});
 }
+
+/**
+ * Handles the payload by filtering out state inputs and returning an array of resolved payload values.
+ * We send null values for state inputs to the server, but we don't want to include them in the resolved payload.
+ *
+ * @param resolved_payload - The resolved payload values received from the client or the server
+ * @param dependency - The dependency object.
+ * @param components - The array of component metadata.
+ * @param with_null_state - Optional. Specifies whether to include null values for state inputs. Default is false.
+ * @returns An array of resolved payload values, filtered based on the dependency and component metadata.
+ */
+export function handle_payload(
+	resolved_payload: unknown[],
+	dependency: Dependency,
+	components: ComponentMeta[],
+	with_null_state = false
+): any[] {
+	try {
+		let payload_index = 0;
+		let updated_payload: unknown[] = [];
+
+		dependency.inputs.forEach((input_id) => {
+			const component = components.find((c) => c.id === input_id);
+			if (component?.type === "state") {
+				if (with_null_state) {
+					updated_payload.push(null);
+				}
+				if (!with_null_state) payload_index++;
+			} else {
+				const value = resolved_payload[payload_index];
+				updated_payload.push(value);
+				payload_index++;
+			}
+		});
+
+		return updated_payload;
+	} catch (e) {
+		console.error(e);
+		return resolved_payload;
+	}
+}
+
+export function build_payload(
+	resolved_payload: unknown[],
+	dependency: Dependency,
+	components: ComponentMeta[]
+): any[] {
+	let updated_payload: unknown[] = [];
+
+	dependency.inputs.map((input_id: number, index: number): any => {
+		const component = components.find((c: any) => c.id === input_id);
+		if (component?.type === "state") {
+			return null;
+		}
+		updated_payload.push(resolved_payload[index]);
+	});
```

Comment:
This is the main issue. That return null isn't being returned anywhere, I'd switch this to a `.forEach` and push the `null` value to the `updated_payload` array as well (or return everything and return the map directly).